test/performance/odp_ipsec.c

```diff
@@ -730,55 +768,76 @@ run_measure_one_async(ipsec_args_t *cargs,
 
 	fill_time_record(start);
 
-	while ((packets_sent < cargs->iteration_count) ||
-	       (packets_received < cargs->iteration_count)) {
+	while ((packets_sent < cargs->packet_count) ||
+	       (packets_received < cargs->packet_count)) {
+		odp_packet_t pkt_out[max_in_flight];
 		odp_event_t ev;
 
-		if ((packets_sent < cargs->iteration_count) &&
-		    (packets_sent - packets_received <
-		     cargs->in_flight)) {
-			pkt = make_packet(pkt_pool, payload_length);
-			if (ODP_PACKET_INVALID == pkt)
+		num_pkts = cargs->packet_count - packets_sent;
+
+		/* Enqueue up to burst size */
+		num_pkts = num_pkts > burst_size ? burst_size : num_pkts;
+
+		/* Enqueue up to (max in flight - current in flight) */
+		in_flight = packets_sent - packets_received;
+		pkts_allowed = max_in_flight - in_flight;
+		num_pkts = num_pkts > pkts_allowed ? pkts_allowed : num_pkts;
```

Comment:
I think this should be:

num_pkts = num_pkts > pkts_allowed ? 0 : num_pkts;

This would make the burst size always (except for the last burst) the same what was specified in the -b option, which is what one would expect. Now the burst size can get smaller if IPsec processing is slower than the test loop (as is expected) and then using a smaller burst size can further slow down the processing so that the loop ends up using small burst size for pretty much the whole test.

And the same comment for the sync case.

With this change (always using the requested burst size), I think the program should complain and exit if the in_flight parameter is less than the burst size parameter.

Reply:
In that case, are you okay with making packet_count a multiple of burst_size? Then we can improve/simplify the run_measure_*() loops.

Reply:
Maybe making packet_count a multiple of burst size does not quite enable the simplification that you may be thinking, because the ipsec out/enq function can still consume only part of the burst, making the count of processed packets not divisible by the burst size.

But simplification is good, and I think the test loop could just stop when the remaining packet count falls below the burst count (or below zero or whatever comes naturally). IOW, the packet count would not be an exact number anymore (because who would actually care?), but just a limit to stop the test at that point.

Actually, I think it would be even better if the test termination would not be based on packet count but elapsed time, as that would make it easier to run the test loop in multiple threads simultaneously without syncing about the number of processed packets. That is what I have been doing with my private test apps (i.e. time based termination & multiple threads running a test loop).

But since this test app does not support multiple threads, packet count limit is ok. I just do not see any need for it to be exact. So a loop that always sends a full burst or nothing and then terminates roughly, but not precisely, after the requested number of packets should be fine. And it would simpler and clearer than having extra logic for handling a partial burst at the end.

Reply:
@JannePeltonen Thanks for your inputs. I hadn't thought about partial enqueue of the burst. So, agreed that such a change won't lead to the simplification I had in mind. So, I'll stick to your earlier suggestion.

Reply:
> With this change (always using the requested burst size), I think the program should complain and exit if the in_flight parameter is less than the burst size parameter.

I think something like this should be done or otherwise the program gets stuck when run with smaller -f than -b.

Reply:
Agreed. I'll have this addressed in v5.